contracts/stake/src/contract.rs

```diff
@@ -35,12 +36,28 @@ impl StakeContract {
         &mut self,
         pk: PublicKey,
         stake: Stake,
+        block_height: BlockHeight,
     ) -> Result<(), Error> {
         let exists = self.staked.get(&pk)?.is_some();
         if exists {
             return Err(Error::StakeAlreadyExists);
         }
 
+        // `created_at` must never be larger than the block height.
+        if stake.created_at() > block_height {
+            return Err(Error::InvalidCreatedAt);
+        }
+
+        // A last_created entry is left when the stake is removed to be able to
+        // make this check. We try to remove it, and if it exists it must not be
+        // larger or equal to the given `created_at`.
+        if let Some(created_at) = self.last_created.remove(&pk)? {
+            if stake.created_at() <= created_at {
+                return Err(Error::InvalidCreatedAt);
+            }
+        }
```

Comment:
Should the previous value be reinserted before returning then error?
Otherwise, if the state is persisted (I'm not sure about it), the check is bypassed calling the stake contract twice in a row

Reply:
This is indeed a problem. There needs to be a true "check" instead of just removing the element. `push_stake` shouldn't corrupt the state if it fails. Even if the call is outside the VM.
Will fix